core/src/main/scala/com/softwaremill/react/kafka_api_opt3/Example.scala

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright (C) 2014 - 2016 Softwaremill <http://softwaremill.com>
+ * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package com.softwaremill.react.kafka_api_opt3
+
+import scala.concurrent.Future
+
+import akka.Done
+import com.softwaremill.react.kafka_api_producer.Producer
+import org.apache.kafka.clients.consumer.KafkaConsumer
+import org.apache.kafka.clients.producer.KafkaProducer
+import org.apache.kafka.clients.producer.ProducerRecord
+
+trait Example {
+  implicit val ec = scala.concurrent.ExecutionContext.Implicits.global
+
+  def consumerProvider: () => KafkaConsumer[Array[Byte], String] = ???
+
+  def producerProvider: () => KafkaProducer[Array[Byte], String] = ???
+
+  class DB {
+    def save(offset: Long, data: String): Future[Done] = ???
+  }
+}
+
+// Consume messages and store a representation, including offset, in DB
+object Example3 extends Example {
+  val db = new DB
+
+  Consumer.source(consumerProvider)
```

Comment:
> Also sliding may hold some elements that will never be emitted at all.

That is a good point! Same is true for `grouped()`.

So then modified:

- stop() is called externally
- Source stops emitting new elements and calls onComplete
- Waits until all messages are committed, or configurable timeout fires
- release the Kafka Consumer

> I dont like put timeout into Source. IMO owner of Source should decide and implement shutdown process. All Source should do is try to shutdown.

I don't see anything wrong with a timeout. We can allow timeout of zero (terminate now, don't care about missing commits) and very long timeouts (wait as long as every commit comes back even if that takes practically forever).

> and in case when messages not committed at all (external offset storing).

It can be done then with timeout of zero then.